### PR TITLE
Add Entra documentation

### LINKED ISSUE #1986: Assigning MFA licenses to users may be a redundant step
## :white_check_mark: Checklist

- [x] I have searched open and closed issues for duplicates
- [x] This is a request for a change to documentation
- [x] This isn't a feature request or bug report
- [x] This isn't an open-ended question (open a [discussion](https://github.com/alan-turing-institute/data-safe-haven/discussions) if it is).

## :ledger: Where is the problem?

[This step](https://data-safe-haven.readthedocs.io/en/latest/roles/system_manager/manage_users.html#assign-mfa-licences) in the docs might be redundant. See discussion in alan-turing-institute/trusted-research#312. As a system manager, I've never had to complete this step.

## :memo: Required documentation changes

Remove this step from the docs if not necessary, or clarify under which circumstances it is required to complete this step.

### LINKED ISSUE #2025: Document Entra directory setup
## :white_check_mark: Checklist

- [x] I have searched open and closed issues for duplicates
- [x] This is a request for a change to documentation
- [x] This isn't a feature request or bug report
- [x] This isn't an open-ended question (open a [discussion](https://github.com/alan-turing-institute/data-safe-haven/discussions) if it is).

## :ledger: Where is the problem?

There is currently no documentation about how to set up the Entra directory

## :memo: Required documentation changes

We should add a new section about setting up Entra, copying relevant content from [previous versions](https://data-safe-haven.readthedocs.io/en/latest/deployment/deploy_shm.html)

### LINKED ISSUE #2049: Users cannot login to workspaces when MFA is required
## :white_check_mark: Checklist

- [x] I have searched open and closed issues for duplicates.
- [x] This is a problem observed when **deploying** a Data Safe Haven.
- [x] I can reproduce this with the [latest version](https://github.com/alan-turing-institute/data-safe-haven/releases).
- [x] I have read through the [documentation](https://alan-turing-institute.github.io/data-safe-haven/).
- [x] This isn't an open-ended question (open a [discussion](https://github.com/alan-turing-institute/data-safe-haven/discussions) if it is).

## :computer: System information

- **Operating System:** macOs
- **Data Safe Haven version:** develop @ 2fd913486

## :package: Packages

<details>
<summary>List of packages</summary>

```none
acme==2.10.0
annotated-types==0.7.0
appdirs==1.4.4
Arpeggio==2.0.2
attrs==23.2.0
azure-common==1.1.28
azure-core==1.30.2
azure-identity==1.17.1
azure-keyvault-certificates==4.8.0
azure-keyvault-keys==4.9.0
azure-keyvault-secrets==4.8.0
azure-mgmt-automation==1.0.0
azure-mgmt-compute==31.0.0
azure-mgmt-containerinstance==10.1.0
azure-mgmt-core==1.4.0
azure-mgmt-dns==8.1.0
azure-mgmt-keyvault==10.3.0
azure-mgmt-msi==7.0.0
azure-mgmt-network==25.4.0
azure-mgmt-rdbms==10.1.0
azure-mgmt-resource==23.1.1
azure-mgmt-storage==21.2.1
azure-storage-blob==12.20.0
azure-storage-file-datalake==12.15.0
azure-storage-file-share==12.16.0
certifi==2024.7.4
cffi==1.16.0
charset-normalizer==3.3.2
chevron==0.14.0
click==8.1.7
cryptography==42.0.8
-e git+https://github.com/craddm/data-safe-haven.git@9e0dd10867265b3270568f9be5a26a57b883df23#egg=data_safe_haven
dill==0.3.8
dnspython==2.6.1
fqdn==1.5.1
grpcio==1.60.1
idna==3.7
isodate==0.6.1
josepy==1.14.0
markdown-it-py==3.0.0
mdurl==0.1.2
msal==1.29.0
msal-extensions==1.2.0
msrest==0.7.1
oauthlib==3.2.2
parver==0.5
portalocker==2.10.1
protobuf==4.25.3
psycopg==3.2.1
pulumi==3.124.0
pulumi_azure_native==2.49.1
pulumi_random==4.16.3
pulumi_tls==5.0.4
pycparser==2.22
pydantic==2.8.2
pydantic_core==2.20.1
Pygments==2.18.0
PyJWT==2.8.0
pyOpenSSL==24.1.0
pyRFC3339==1.1
pytz==2024.1
PyYAML==6.0.1
requests==2.32.3
requests-oauthlib==2.0.0
rich==13.7.1
semver==2.13.0
setuptools==70.3.0
shellingham==1.5.4
simple_acme_dns==3.1.0
six==1.16.0
typer==0.12.3
typing_extensions==4.12.2
urllib3==2.2.2
validators==0.28.3
websocket-client==1.8.0
```

</details>

## :no_entry_sign: Describe the problem

Users cannot login to workspaces within SREs. The error message typically means the password is incorrect, but it is correct. An admin account can login at the serial console. 

`sudo password -S <user.name>` indicates the passwords for the users are locked.

Update:

The problem is with the Require MFA Conditional Access Policy. When trying to login to the Ubuntu desktop, a request is made for MFA authentication (presumably because it's coming from a different IP address than you've already logged in to guacamole from) which never reaches the user.

## :deciduous_tree: Log messages

<details>
<summary>Relevant log messages</summary>

image

</details>

## :recycle: To reproduce

Deploy an SRE attached to an Entra ID with the Require MFA Conditional Access policy set up and attempt to login to the workspace.

### PR DESCRIPTION
**Pull request by jemrobinson**
## :white_check_mark: Checklist

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [x] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).

### :vertical_traffic_light: Depends on

n/a

### :arrow_heading_up: Summary

Document how to set up an Entra directory

**Note** We want our conditional access policy to apply to `sre-example-guacamole` but not to `sre-example-apricot`. The only attribute we seem to be able to successfully filter on is "Operating System" which shows up as `null` for the `apricot` login. This may need to be revisited in future - ideally Entra will fix the ability to filter on ApplicationID so that it also applies to "Desktop" apps as well as "WebApps".

Screenshot 2024-07-26 at 06 51 10

### :closed_umbrella: Related issues

Closes #1986
Closes #2025
Closes #2049

### :microscope: Tests

Documentation only. Tested the CA rules work on a pre-deployed SRE.

### CODE REVIEW

**Comment by jemrobinson** [comment]
**Note** We can change the default sign-in frequency from once every 90 days. Would once every day make sense?

**Comment by JimMadge** [comment]
> **Note** We can change the default sign-in frequency from once every 90 days. Would once every day make sense?

That sounds sensible to me. Not frequent enough to be annoying. Frequent enough to make an unattended computer less of a risk.

**Comment by jemrobinson** [comment]
Screenshot 2024-07-26 at 09 36 35

Another option is "Every time"

> "Every time" option is evaluated on every sign-in attempt to an application in scope for this policy. Some policy configurations for the "sign-in frequency every time" session control are in preview.

but this might be annoying. Thoughts @JimMadge @craddm ?

File: docs/source/deployment/configure_entra_id.md [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+If you do not already have access to a **native** administrator account, create one using the steps below.
+
+:::{admonition} How to create a native Entra administrator
+:class: dropdown hint
+Follow the instructions [here](https://learn.microsoft.com/en-us/entra/fundamentals/how-to-create-delete-users#create-a-new-user).
+Use the following settings:
+
+- `Basics` tab:
+    - **User principal name:** `entra.admin.firstname.lastname`
+        - If you have a choice of domains use `<your organisation>.onmicrosoft.com` which will create a clearer separation between administrators and users
```
Comment (JimMadge) [inline docs/source/deployment/configure_entra_id.md]:
```suggestion
        - If you have a choice of domains use `<your organisation>.onmicrosoft.com`, which will create a clearer separation between administrators and users
```

File: docs/source/deployment/configure_entra_id.md [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+If you created a new Microsoft Entra tenant, an external administrator account will have been automatically created for you.
+If you do not already have access to a **native** administrator account, create one using the steps below.
+
+:::{admonition} How to create a native Entra administrator
+:class: dropdown hint
+Follow the instructions [here](https://learn.microsoft.com/en-us/entra/fundamentals/how-to-create-delete-users#create-a-new-user).
+Use the following settings:
+
+- `Basics` tab:
+    - **User principal name:** `entra.admin.firstname.lastname`
```
Comment (JimMadge) [inline docs/source/deployment/configure_entra_id.md]:
```suggestion
    - **User principal name:** `entra.admin.<firstname>.<lastname>`
```

File: docs/source/deployment/configure_entra_id.md [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        - If you have a choice of domains use `<your organisation>.onmicrosoft.com` which will create a clearer separation between administrators and users
+    - **Display name:** `Entra Admin - Firstname Lastname`
```
Comment (JimMadge) [inline docs/source/deployment/configure_entra_id.md]:
```suggestion
    - **Display name:** `Entra Admin - <Firstname> <Lastname>`
```

File: docs/source/deployment/configure_entra_id.md [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+:::
+
+## Register allowed authentication methods
+
+In this section, you will determine which methods are permitted for multi-factor authentication (MFA).
+This is necessary both to secure logins and to allow users to set their own passwords.
+
+- Sign in to the [Microsoft Entra admin centre](https://entra.microsoft.com/)
+- Browse to **Protection > Authentication methods** from the menu on the left side.
+- Click **Manage > Policies** on the internal menu on the left side.
```
Comment (JimMadge) [inline docs/source/deployment/configure_entra_id.md]:
```suggestion
- Browse to **Protection > Authentication methods** from the menu on the left side
- Click **Manage > Policies** on the internal menu on the left side
```

File: docs/source/deployment/configure_entra_id.md [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+- Ensure that you provide **both** a phone number **and** an email address.
+
+:::
+
+Now you can reset the password for this user, following the steps below.
+
+:::{admonition} How to reset your Entra user password
+:class: dropdown hint
+
+- Go [here](https://passwordreset.microsoftonline.com/) and follow the instructions to set your password.
```
Comment (JimMadge) [inline docs/source/deployment/configure_entra_id.md]:
```suggestion
- Follow the instructions [here](https://passwordreset.microsoftonline.com/) to set your password
```

File: docs/source/deployment/configure_entra_id.md [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+- You will need access to the phone number and/or email address from the previous step.
```
Comment (JimMadge) [inline docs/source/deployment/configure_entra_id.md]:
```suggestion
- You will need access to the phone number and/or email address from the previous step
```

File: docs/source/deployment/configure_entra_id.md [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+:::
+
+## Delete any external administrators
+
+:::{warning}
+In this step we will delete any external admin account which might belong to Microsoft Entra ID.
+Before you do this, you **must** ensure that you can log into Entra using your **native** administrator account.
+:::
+
+Start by identifying whether you have any external users
```
Comment (JimMadge) [inline docs/source/deployment/configure_entra_id.md]:
```suggestion
Start by identifying whether you have any external users.
```

File: docs/source/deployment/configure_entra_id.md [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+To do so, follow the instructions [here](https://learn.microsoft.com/en-us/entra/identity/role-based-access-control/security-emergency-access).
+Since this account will be exempt from normal login policies, it should not be used except when **absolutely necessary**.
+:::
+
+## Purchase Microsoft Entra licences
+
+At least one user needs to have a [Microsoft Entra Licence](https://www.microsoft.com/en-gb/security/business/microsoft-entra-pricing) assigned in order to enable [self-service password reset](https://learn.microsoft.com/en-us/entra/identity/authentication/concept-sspr-licensing) and conditional access policies.
+
+:::{tip}
+P1 Licences are sufficient, but you may use another licence if you prefer.
```
Comment (JimMadge) [inline docs/source/deployment/configure_entra_id.md]:
```suggestion
P1 Licences are sufficient but you may use another licence if you prefer.
```

File: docs/source/deployment/configure_entra_id.md [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+:::
+
+- Sign in to the [Microsoft Entra admin centre](https://entra.microsoft.com/)
+- Browse to **Identity > Billing > Licenses** from the menu on the left side.
+- Click on `All products` on the internal menu on the left side.
```
Comment (JimMadge) [inline docs/source/deployment/configure_entra_id.md]:
```suggestion
- Browse to **Identity > Billing > Licenses** from the menu on the left side
- Click on `All products` on the internal menu on the left side
```

File: docs/source/deployment/configure_entra_id.md [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+    - Wait for the selected licence to appear on the `All products` list (this may take several minutes)
+
+## Enable self-service password reset
+
+In order to enable self-service password reset (SSPR) you will need to do the following:
+
+- Sign in to the [Microsoft Entra admin centre](https://entra.microsoft.com/)
+- Browse to **Protection > Password reset** from the menu on the left side.
+- Click **Manage > Properties** on the internal menu on the left side.
+- Under the option `Self service password reset enabled`, choose **All**.
```
Comment (JimMadge) [inline docs/source/deployment/configure_entra_id.md]:
```suggestion
- Browse to **Protection > Password reset** from the menu on the left side
- Click **Manage > Properties** on the internal menu on the left side
- Under the option `Self service password reset enabled`, choose **All**
```

File: docs/source/deployment/configure_entra_id.md [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+
+## Disable security defaults
+
+- Sign in to the [Microsoft Entra admin centre](https://entra.microsoft.com/)
+- Browse to **Identity > Overview > Properties** from the menu on the left side.
```
Comment (JimMadge) [inline docs/source/deployment/configure_entra_id.md]:
```suggestion
- Browse to **Identity > Overview > Properties** from the menu on the left side
```

File: docs/source/deployment/configure_entra_id.md [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+    - **Security defaults** to `Disabled (not recommended)`
+    - Select **My organization is planning to use Conditional Access**
+    - Click the `Save` button
+- At the prompt click the `Disable` button
+
+## Apply conditional access policies
+
+- Sign in to the [Microsoft Entra admin centre](https://entra.microsoft.com/)
+- Browse to **Protection > Conditional Access** from the menu on the left side.
+- Click **Policies** on the internal menu on the left side.
```
Comment (JimMadge) [inline docs/source/deployment/configure_entra_id.md]:
```suggestion
- Browse to **Protection > Conditional Access** from the menu on the left side
- Click **Policies** on the internal menu on the left side
```

File: docs/source/deployment/configure_entra_id.md [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+- Click the `Create` button
+
+### Restrict Microsoft Entra ID access
+
+- Create a new policy named `Restrict Microsoft Entra ID access`
+- Under `Users` set:
+    - **Include**: Select `All users`
+    - **Exclude**:
+        - Check `Directory roles`
+        - In the drop-down menu select `Global administrator`.
```
Comment (JimMadge) [inline docs/source/deployment/configure_entra_id.md]:
```suggestion
        - In the drop-down menu select `Global administrator`
```

File: docs/source/deployment/configure_entra_id.md [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+- At the prompt click the `Disable` button
+
+## Apply conditional access policies
+
+- Sign in to the [Microsoft Entra admin centre](https://entra.microsoft.com/)
+- Browse to **Protection > Conditional Access** from the menu on the left side.
+- Click **Policies** on the internal menu on the left side.
+
+### Require MFA
+
```
Comment (JimMadge) [inline docs/source/deployment/configure_entra_id.md:145]:
```suggestion
### Require MFA

These instructions will create a policy which requires all users (except the emergency administrator if you have created one) to use multi-factor authentication (MFA) to log in.

```

File: docs/source/deployment/configure_entra_id.md [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+- Under `Session`:
+    - Check `Sign-in frequency`
+    - Check `Periodic reauthentication`
+        - Set the value to `1 day(s)`
+- Under `Enable policy` select `On`
+    - Check `I understand that my account will be impacted by this policy. Proceed anyway.`
+- Click the `Create` button
+
+### Restrict Microsoft Entra ID access
+
```
Comment (JimMadge) [inline docs/source/deployment/configure_entra_id.md:180]:
```suggestion
### Restrict Microsoft Entra ID access

These instructions will prevent non-administrators from being able to view the Entra ID configuration.

```

File: docs/source/management/index.md [hunk 16]
```diff
@@ -47,6 +47,23 @@ Usernames are of the format `<GivenName>.<Surname>` and do not include the Entra
 $ dsh users list <SRE name>
 ```
 
+## Manually register users for SSPR
+
+:::{tip}
+Users created via the `dsh users` command line tool will be automatically registered for SSPR
```
Comment (JimMadge) [inline docs/source/management/index.md]:
```suggestion
Users created via the `dsh users` command line tool will be automatically registered for SSPR.
```

File: docs/source/management/index.md [hunk 17]
```diff
@@ -47,6 +47,23 @@ Usernames are of the format `<GivenName>.<Surname>` and do not include the Entra
 $ dsh users list <SRE name>
 ```
 
+## Manually register users for SSPR
```
Comment (JimMadge) [inline docs/source/management/index.md]: Is SSPR defined?

File: docs/source/management/index.md [hunk 18]
```diff
@@ -47,6 +47,23 @@ Usernames are of the format `<GivenName>.<Surname>` and do not include the Entra
...
+Users created via the `dsh users` command line tool will be automatically registered for SSPR
+:::
+
+If you have manually created a user and want to enable SSPR, do the following
+
+- Go to the [Microsoft Entra admin centre](https://entra.microsoft.com/)
+- Browse to **Users > All Users** from the menu on the left side
+- Select the user you want to enable SSPR for
+- On the **Manage > Authentication Methods** page fill out their contact info as follows:
+    - Phone: add the user's phone number with a space between the country code and the rest of the number (e.g. `+44 7700900000`).
```
Comment (JimMadge) [inline docs/source/management/index.md]:
```suggestion
    - Phone: add the user's phone number with a space between the country code and the rest of the number (_e.g._ `+44 7700900000`)
```

File: docs/source/management/index.md [hunk 19]
```diff
@@ -47,6 +47,23 @@ Usernames are of the format `<GivenName>.<Surname>` and do not include the Entra
...
+    - Email: enter the user's email address here
+    - Ensure that you have registered **both** a phone number and an email address.
```
Comment (JimMadge) [inline docs/source/management/index.md]:
```suggestion
    - Ensure that you have registered **both** a phone number and an email address
```

**Review by JimMadge — CHANGES_REQUESTED** [review]
Looks good.

A few style changes and places to add some clarifying text.

I don't understand the usage of fixed width and bold, it feels inconsistent.
Do we have a place where we have said how/why we use them?

My preference would be to use emphasis over bold, bold sparingly and only use fixed width for code/variables/expressions.

**Comment by JimMadge** [comment]
@jemrobinson I think every time would be annoying.

Perhaps this is something we want to leave open to users though. I could see it becoming appropriate for more sensitive information.

**Comment by JimMadge** [comment]
These dropdown hints look great :+1:

**Comment by jemrobinson** [comment]
> I don't understand the usage of fixed width and bold, it feels inconsistent.
> Do we have a place where we have said how/why we use them?

The Microsoft Entra docs use bold and > to indicate navigation through menus which I've tried to use here for consistency. It would be nice to use something (possibly fixed width?) to indicate things we want the user to select/type, but I don't know if I've been consistent with that. If you have a suggested style, I'm happy to change this PR to meet it.

**Comment by JimMadge** [comment]
I think what bothers me the most is the inconsistently in **click me** and `click me`. How about,

- **bold** when we want to draw attention inline, user navigation or clicks
- *emph* foreign words and abbreviations
- `fixed width` code, expressions that the user types (`ls -l <path/to/directory>`), variable names

It will be easier to change later (if we want to) if we are as consistent as possible now.

**Comment by jemrobinson** [comment]
[Google recommend](https://developers.google.com/style/placeholders) using uppercase characters with underscore delimiters. (*LIKE_THIS*) for text that you want the user to replace. I might try that in addition to your suggestions.

File: docs/source/management/index.md (hunk 17 above)
Comment (jemrobinson) [reply docs/source/management/index.md]:
```suggestion
## Manually register users for self-service password reset
```

**Comment by jemrobinson** [comment]
Thanks for these suggestions @JimMadge : I think it already looks much better.

**Review by JimMadge — APPROVED** [review]

**Comment by jemrobinson** [comment]
Just found out we can have semantic `menuselection` and `guilabel` attributes so I've added these.

**Pull request merged**